src/main/java/com/google/firebase/auth/hash/Argon2.java

```diff
@@ -0,0 +1,153 @@
+    checkArgument(builder.parallelism >= 1 && builder.parallelism <= 16,
+        "parallelism is required for Argon2 and must be between 1 and 16");
+    checkArgument(builder.iterations >= 1 && builder.iterations <= 16,
+        "iterations is required for Argon2 and must be between 1 and 16");
+    checkArgument(builder.memoryCostKib > 0 && builder.memoryCostKib <= 32768,
+        "memoryCostKib is required for Argon2 and must be less than or equal to 32768");
+    this.hashLengthBytes = builder.hashLengthBytes;
+    this.hashType = builder.hashType;
+    this.parallelism = builder.parallelism;
+    this.iterations = builder.iterations;
+    this.memoryCostKib = builder.memoryCostKib;
+    if (builder.version != null) {
+      this.version = builder.version;
+    } else {
+      this.version = null;
+    }
+    if (builder.associatedData != null) {
+      this.associatedData = BaseEncoding.base64Url().encode(builder.associatedData);
+    } else {
+      this.associatedData = null;
+    }
+  }
+
+  @Override
+  protected Map<String, Object> getOptions() {
+    ImmutableMap.Builder<String, Object> builder = ImmutableMap.<String, Object>builder()
+        .put("hashLengthBytes", hashLengthBytes)
+        .put("hashType", hashType.toString())
+        .put("parallelism", parallelism)
+        .put("iterations", iterations)
+        .put("memoryCostKib", memoryCostKib);
+    if (this.associatedData != null) {
+      builder.put("associatedData", associatedData);
+    }
+    if (this.version != null) {
+      builder.put("version", version.toString());
+    }
+    return builder.build();
+  }
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  public static class Builder {
+
+    private int hashLengthBytes;
+    private Argon2HashType hashType;
+    private int parallelism;
+    private int iterations;
+    private int memoryCostKib;
+    private Argon2Version version;
+    private byte[] associatedData;
+
+    private Builder() {
+    }
+
+    public Builder setHashLengthBytes(int hashLengthBytes) {
+      this.hashLengthBytes = hashLengthBytes;
+      return this;
+    }
+
+    public Builder setHashType(Argon2HashType hashType) {
+      this.hashType = hashType;
+      return this;
+    }
+
+    public Builder setParallelism(int parallelism) {
+      this.parallelism = parallelism;
+      return this;
+    }
+
+    public Builder setIterations(int iterations) {
+      this.iterations = iterations;
+      return this;
+    }
+
+    public Builder setMemoryCostKib(int memoryCostKib) {
+      this.memoryCostKib = memoryCostKib;
+      return this;
+    }
+
+    public Builder setVersion(Argon2Version version) {
+      this.version = version;
+      return this;
+    }
+
+    public Builder setAssociatedData(byte[] associatedData) {
+      this.associatedData = associatedData;
+      return this;
+    }
+
+    public Argon2 build() {
+      return new Argon2(this);
+    }
+  }
+
+  public enum Argon2HashType {
```

Comment:
I wonder if this could use another layer of abstraction. Similar to what we have done with `Hmac` and `RepeatableHash`. We could introduce an abstract class for `ARGON2Hash` and implementations for `ARGON2d`, `ARGON2i`, and `ARGON2id`. WDYT?

Reply:
I think that's unnecessary here. For HMAC, it makes sense to add a layer of abstraction since the algorithms are identified as different from one another (`HMAC_SHA256`  `HMAC_SHA1` `HMAC_MD5`, etc). Here `ARGON2` is the algorithm, the hash-types are an additional configuration for the same algorithm.

Reply:
I see. That makes sense. Thanks, Sam!